File: 2019/day2.js
const f = function(x) {
  const xs = x.split(",").map(a => parseInt(a));

  for (p = 0; ; p += 4) {
    if (xs[p] != null) {
      //console.log(xs[p]);
      if (xs[p] == 1) {
        //console.log("sum");
        xs[xs[p+3]] = xs[xs[p+1]] + xs[xs[p+2]];
      } else if (xs[p] == 2) {
        xs[xs[p+3]] = xs[xs[p+1]] * xs[xs[p+2]];
      } else if (xs[p] == 99) {
        break;
      } else {
        throw "not expected!";
      }
    } else {
      break;
    }
  }
  return xs.toString();
}

const v = function(x, y, z) {
  const b = x.split(",");
  b[1] = y;
  b[2] = z;
  return b.toString();
}

const g = function(x) {
  const b = x.split(",");
  return parseInt(b[0]);
}

const input = "1,69,79,3,1,1,2,3,1,3,4,3,1,5,0,3,2,1,6,19,1,9,19,23,2,23,10,27,1,27,5,31,1,31,6,35,1,6,35,39,2,39,13,43,1,9,43,47,2,9,47,51,1,51,6,55,2,55,10,59,1,59,5,63,2,10,63,67,2,9,67,71,1,71,5,75,2,10,75,79,1,79,6,83,2,10,83,87,1,5,87,91,2,9,91,95,1,95,5,99,1,99,2,103,1,103,13,0,99,2,14,0,0";
const aim = 19690720;

console.log(f(input));

let a = 0;
for (i = 0; i <= 99; i++) {
  for (j = 0; j <= 99; j++) {
    a = f(v(input, i, j));
    if (g(a) == aim) {
      console.log((i*100)+j);
      break;
    }
  }
  if (g(a) == aim) break;
}

File: 2023/AoC2023/day2.fs
module Day2

open Utils

let partTwo = true

type Round = {
    Red: int
    Green: int
    Blue: int
}

type Game = {
    Id: int
    Rounds: Round array
}

let maxForGame (game: Game) : Round =
    game.Rounds
    |> Array.fold (fun acc curr -> { 
        Red = if curr.Red > acc.Red then curr.Red else acc.Red
        Green = if curr.Green > acc.Green then curr.Green else acc.Green
        Blue = if curr.Blue > acc.Blue then curr.Blue else acc.Blue }) { Red = 0; Green = 0; Blue = 0 }

let product (round: Round): int =
    round.Red * round.Green * round.Blue

let roundWouldBePossible (max: Round) (round: Round)  =
    round.Red <= max.Red && round.Blue <= max.Blue && round.Green <= max.Green

let gameWouldBePossible (game: Game) =
    let b = 
        game.Rounds 
        |> Array.map (roundWouldBePossible { Red = 12; Blue = 14; Green = 13 })
        |> Array.contains false
    not b

let getColour (section: string) (colour: string) =
    let colourSection = 
        section.Split ", "
        |> Array.filter (fun x -> x.Contains colour)
        |> (fun x -> if Array.length x = 0 then "" else Array.head x)
    if colourSection = "" then 0 else colourSection.Split " " |> Array.head |> int

let gameId (game: Game) = game.Id

let getRound section =
    { Red = getColour section "red"; Green = getColour section "green"; Blue = getColour section "blue"; }
    

let getRounds (section: string) =
    section.Split "; "
    |> Array.map getRound

let parseLine (line: string): Game =
    let colon = line.Split ": "
    let gameId = int (colon[0].Substring 5)
    { Id = gameId; Rounds = getRounds colon[1] }

let runDay (input: string) =
    let parsed =
        input
        |> splitByLine
        |> Array.map parseLine
    if partTwo
        then
            parsed
            |> Array.map maxForGame
            |> Array.map product
            |> Array.sum
        else
            parsed
            |> Array.filter gameWouldBePossible
            |> Array.map gameId
            |> Array.sum
